### ISSUE TITLE
Memory consumption after rendering certain amount of pdfs

### ISSUE THREAD

**Issue by quarryman** [wojtekmaj/react-pdf#305]
**Before you start - checklist**

 - [x ] I followed instructions in documentation written for my React-PDF version
 - [x ] I have checked if this bug is not already reported
 - [x ] I have checked if an issue is not listed in [Known issues](https://github.com/wojtekmaj/react-pdf/wiki/Known-issues)

**Description**

After rendering certain amount of pdfs we see memory consumption is constantly growing and memory is never released.

According to chrome profiler most memory usage is due to 

```MessageHandler._onComObjOnMessage method```

Here is a screenshot of profiling results

![screen shot 2018-11-15 at 14 19 55](https://user-images.githubusercontent.com/1093040/48977463-3e475d80-f0a3-11e8-8d1f-b7f6324c44a4.png)

Its probably related to pdf.js itself, so first I filed an issue on pfd.js project.

[initial pdf.js issue](https://github.com/mozilla/pdf.js/issues/10260)

It was reviewed and the result is that they dont consider this memory leak, unless react-pdf implementation doesn't invoke 'destroy' method for 'loadingTask'

https://github.com/mozilla/pdf.js/blob/960213cb6963f9bf46d6224805de2dc73fef0af4/src/display/api.js#L481-L498

And I see destroy for loadingTask is never invoked.

Here was my question to pdf.js

> Is it right that we MUST call destroy() to avoid memory consumption and if we don't it may lead to situation above?

And here is pdf.js guys answer to it

>The loadingTask must be destroy()-ed in order to clean-up various resources, both on the main- and worker-threads.

I tried brief implementation of 'destroy' this way

```javascript
// utils.js
export const makeCancellable = (promise) => {
  let isCancelled = false;

  const wrappedPromise = new Promise((resolve, reject) => {
    promise.then(
      (...args) => (isCancelled ? reject(new PromiseCancelledException('Promise cancelled')) : resolve(...args)),
      error => (isCancelled ? reject(new PromiseCancelledException('Promise cancelled')) : reject(error)),
    );
  });

  return {
    promise: wrappedPromise,
    cancel() {
      isCancelled = true;
    },
    destroy() {
      if (typeof promise.destroy === 'function') {
        promise.destroy();
      }
    },
  };
};

// Document,jsx
componentWillUnmount() {
    this.runningTask.destroy();
    cancelRunningTask(this.runningTask);
  }
```
Its probably  playing around of promises which is not right way to handle it, but anyway, it leads to no memory consumption,

```promise``` is an instance of ```LoadingTask``` (PDFJS.getDocument(source)) and ``` promise.destroy() ``` is successfully invoked and releases resources,  but is causing this error from time to time and i think its due my misunderstanding when is must be invoked.

```javascript
pdf.worker.js:22726 Uncaught (in promise) Error: Worker was terminated
    at ensureNotTerminated (pdf.worker.js:22726)
    at onFailure (pdf.worker.js:22866)
```

I have selenium long running test, that shows memory is released with destroy method invoked, and is not released without it.

I think I dont understand the right pdf rendering flow to get loadingTask destroyed when it must be, but I believe this must be invoked as guys from pdf.js confirm this.
Can we try clarifying the right flow - when loading task can be destroyed and is it finally the right approach to free resources.

```this.runningTask = makeCancellable(PDFJS.getDocument(source));```

this is pdf.js API wrapped into cancellable promise, and this API has 'destroy' method, that I referenced with a link earlier and that we never invoke.

**Steps to reproduce**

1. Render pdf document with Document component.
2. While rendering has not yet finished, unmount Document component and mount it again with new pdf source. 
3. See memory consumption

**Expected behavior**

Expect to see memory released

**Additional information**

**Environment**
 - Chrome 70.0
 - React-PDF version 3.0.5
 - React version  16.2.0

**Comment by quarryman** [comment]
@wojtekmaj hello.
Are there any chances you will take a look at the issue or leave your thoughts on the problem at least?
Thanx much

**Comment by wojtekmaj** [comment]
Hi there, thanks for reporting this. Could you check if branch `destroy-task` on which I included the potential fix, behaves any better for you?

**Comment by quarryman** [comment]
@wojtekmaj I will try within a few hours and give our feedback.
Thanx much for participating! We very appreciate your help.

**Comment by wojtekmaj** [comment]
You shouldn't create these file objects on the fly in `render` function, but the assumption that everyone would follow this rule was stupid of me.

If I can be of any help during upgrade, let me know. There's also an upgrade guide on Wiki.

**Comment by quarryman** [comment]
@wojtekmaj hello
It takes me some time to investigate, so sorry for slow updates on the issue.

> You shouldn't create these file objects on the fly in render function

I just pass 'file' property to Document component as object, as I need to pass password for protected documents like this
```
<Document
        file={{ url, password }}
        inputRef={this.handleRef}
 >
...
</Document>
```
Do you mean the right way is to create object before render and just pass referense so that 2 objects comparison will see its the same object like this?

```
<Document
        file={this.fileObj}
        inputRef={this.handleRef}
 >
...
</Document>

Another problems i see with destroy implementation is that I constantly see react warning Document component trying to call setState on an unmounted component and it fails to render pdf from time to time.
Here is the reason I think may cause this in both v4 and v3.
this.runningTask is cancellable promise and its cancelled in ComponentWillUnmount, but 'this.loadingTask' is not (as far as I understand)

Given we have next react flow:
ComponentDidUpdate
ComponentWillUnmount

Once ComponentDidUpdate is finished it has this.loadingTask.destroy() promise in pending state.
Once ComponentWillUnmount is finished it doesnt cancel loadingTask.destroy(), so destroy resolves and invokes this.loadDocument() that then calls setState on unmounted component.

**Comment by nikhilag** [comment]
@quarryman Were you able to resolve this issue?

**Comment by mfferreira** [comment]
bump

**Comment by wojtekmaj** [comment]
Duplicate of #288

**Comment by wojtekmaj** [comment]
Please kindly check React-PDF [`v5.3.0-beta.2`](https://github.com/wojtekmaj/react-pdf/releases/tag/v5.3.0-beta.2), in which improvements regarding memory management were made.

Let me know what you think in #748!
